src/chunk.c

```diff
@@ -3210,43 +3210,68 @@ ts_chunk_drop_process_materialization(Oid hypertable_relid,
 	ts_cache_release(hcache);
 }
 
+static void
+lock_referenced_tables(Oid table_relid)
+{
+	List *fk_relids = NIL;
+	ListCell *lf;
+	List *cachedfkeys = NIL;
+	Relation table_rel = table_open(table_relid, AccessShareLock);
+
+	/* this list is from the relcache and can disappear with a cache flush, so
+	 * no further catalog access till we save the fk relids */
+	cachedfkeys = RelationGetFKeyList(table_rel);
+	foreach (lf, cachedfkeys)
+	{
+		ForeignKeyCacheInfo *cachedfk = lfirst_node(ForeignKeyCacheInfo, lf);
+
+		/* conrelid should always be that of the table we're considering */
+		Assert(cachedfk->conrelid == RelationGetRelid(table_rel));
+		fk_relids = lappend_oid(fk_relids, cachedfk->confrelid);
+	}
+	table_close(table_rel, AccessShareLock);
+	foreach (lf, fk_relids)
+		LockRelationOid(lfirst_oid(lf), AccessExclusiveLock);
+}
+
 /* Continuous agg materialization hypertables can be dropped
  * only if a user explicitly specifies the table name
  */
 List *
-ts_chunk_do_drop_chunks(Oid table_relid, Datum older_than_datum, Datum newer_than_datum,
+ts_chunk_do_drop_chunks(Hypertable *ht, Datum older_than_datum, Datum newer_than_datum,
 						Oid older_than_type, Oid newer_than_type,
 						CascadeToMaterializationOption cascades_to_materializations,
-						int32 log_level, bool user_supplied_table_name, List **affected_data_nodes)
+						int32 log_level, List **affected_data_nodes)
 
 {
 	uint64 i = 0;
 	uint64 num_chunks = 0;
 	Chunk *chunks;
 	List *dropped_chunk_names = NIL;
 	const char *schema_name, *table_name;
-	int32 hypertable_id = ts_hypertable_relid_to_id(table_relid);
+	const int32 hypertable_id = ht->fd.id;
 	bool has_continuous_aggs;
 	List *data_nodes = NIL;
 
-	Assert(OidIsValid(table_relid));
+	ts_hypertable_permissions_check(ht->main_table_relid, GetUserId());
 
-	ts_hypertable_permissions_check(table_relid, GetUserId());
+	/* We have a FK between hypertable H and PAR. Hypertable H has number of
+	 * chunks C1, C2, etc. When we execute "drop table C", PG acquires locks
+	 * on C and PAR. If we have a query as "select * from hypertable", this
+	 * acquires a lock on C and PAR as well. But the order of the locks is not
+	 * the same and results in deadlocks. - github issue 865 We hope to
```

Comment:
If you don't mind to make the following update, so GitHub introduces hyperlink to the issue:
```suggestion
	 * the same and results in deadlocks. - github issue #865 We hope to
```